packages/block-library/src/paragraph/controls.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * WordPress dependencies
+ */
+import { __, _x, isRTL } from '@wordpress/i18n';
+import {
+	ToolbarButton,
+	ToggleControl,
+	__experimentalToolsPanelItem as ToolsPanelItem,
+} from '@wordpress/components';
+import {
+	AlignmentControl,
+	BlockControls,
+	InspectorControls,
+	useSettings,
+	store as blockEditorStore,
+} from '@wordpress/block-editor';
+import { formatLtr } from '@wordpress/icons';
+import { useSelect } from '@wordpress/data';
+import { pure } from '@wordpress/compose';
+
+function ParagraphRTLControl( { direction, setDirection } ) {
+	return (
+		isRTL() && (
+			<ToolbarButton
+				icon={ formatLtr }
+				title={ _x( 'Left to right', 'editor button' ) }
+				isActive={ direction === 'ltr' }
+				onClick={ () => {
+					setDirection( direction === 'ltr' ? undefined : 'ltr' );
+				} }
+			/>
+		)
+	);
+}
+
+export function hasDropCapDisabled( align ) {
+	return align === ( isRTL() ? 'left' : 'right' ) || align === 'center';
+}
+
+function Controls( { clientId, setAttributes } ) {
```

Comment:
The main gotcha here is that this component doesn't depend on attributes and rely on useSelect to fetch the attributes instead.

I wonder if it's better to continue relying on attributes (because the wrapper will rerender anyway, but have three pure components (one of align, one for drop cap and one for rtl) which would avoid the useSelect

Now, I wonder about the feeling of over-optimization. So I'm kind of curious about the numbers of such refactor in general.

Also, I think we should draw some lesson from this PR: 

 - Either tell people to have "pure components" per control.
 - Or normalize a controls.js + useSelect file per block. 

I'm wondering also if the controls.js approach only works in the paragraph component because we don't access the "content" within the controls and that it might not be the case in all blocks which means the useSelect will be just an overhead instead of an optimization.

Reply:
Yes, it may be better for now to continue relying on the attributes as props, and only pass the needed ones. That's actually what I have done for now in #56813.

I don't think it's an over-optimisation if there's a clear difference in the numbers. Then most other performance PRs are over-optimisations. :D 

Yes, as noted in the description, I do wonder what that means for the API. Tbh, I'm never been a big fan of having the controls and the content in the same Edit function. But yes, if we keep that it might be better to suggest making "pure" controls.

I don't think we access the content in many controls at all, if any at all.

Reply:
Actually, this PR is going to break "Block Bindings". Block Bindings work by updating the "attributes" prop that get injected into BlockEdit and this PR bypass that. So I'm thinking we should not merge it and instead do the granular controls approach.

Reply:
Can you explain this a bit more? What are "Block Bindings"? You mean mutating the attributes object?

Reply:
I don't find anything related to bindings for blocks in the codebase.